### ISSUE TITLE
Walktrap community detection fails on some disconnected weighted graphs

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2042]
**Describe the bug**

`igraph_community_walktrap()` fails on some disconnected weighted graphs.

This was reported here: https://igraph.discourse.group/t/number-of-steps-is-greater-than-number-of-rows-in-merges-matrix/1184?u=szhorvat

**To reproduce**

In Mathematica,

```
pg = IGShorthand["1,2-4,2-5,3-5,4-5", 
  EdgeWeight -> {0.2, 0.2, 0.1, 0.4}, VertexLabels -> Automatic, 
  EdgeLabels -> "EdgeWeight"]
```

<img width="360" alt="image" src="https://user-images.githubusercontent.com/1212871/163809010-f53ba1c2-2352-4a4f-9543-e6f6a7889704.png">

```
In[285]:= IGCommunitiesWalktrap[pg]

During evaluation of In[285]:= IGraphM::error: src/community/community_misc.c:112 - Number of steps is greater than number of rows in merges matrix: found 4 steps, 3 rows.

During evaluation of In[285]:= IGraphM::error: igraph returned with error: Invalid value.

Out[285]= $Failed
```

I found this example with fuzzing, then manual tweaking of the weights for a simpler output. Original fuzzing code:

```
Do[g = RandomGraph[BernoulliGraphDistribution[5, 0.3], 
   EdgeWeight -> {_ :> RandomReal[{-.1, 0.5}]}];
 Check[IGCommunitiesWalktrap[g], Abort[]],
 Infinity]
g
IGEdgeProp[EdgeWeight][g]
```

(However, negative weights are not required.)

I cannot find examples with all-integer weights, or with no weights. I tried the first 1000 graphs from the graph atlas, and none failed.

**Version information**

Current `master` branch (i.e. soon after 0.9.8).

**Comment by szhorvat** [comment]
A simpler example is a graph with three vertices, where only two are connected, and the weight is 0.2. Not every weight will reproduce the problem.

In R,

```
> g<-graph_from_literal(1-2,3)
> E(g)$weight <- c(0.2)
> cluster_walktrap(g)
Error in cluster_walktrap(g) : 
  At core/community/community_misc.c:112 : Number of steps is greater than number of rows in merges matrix: found 2 steps, 1 rows. Invalid value
```

**Comment by szhorvat** [comment]
This might have to do with the fact that the `merges` and `modularity` vectors are not of the same size. I believe they should be.

Example again in Mathematica:

```
In[391]:= cl = IGCommunitiesWalktrap[IGEmptyGraph[5]];
cl["Merges"]
cl["Modularity"]

Out[392]= {}
Out[393]= {0., 0., 0., 0., 0.}
```

And in R:

```
> cl <- cluster_walktrap(make_empty_graph(5, directed = F))
> cl$merges
     [,1] [,2]
> cl$modularity
[1] 0 0 0 0 0
```

This is a graph with 5 vertices and no edges. The `merges` matrix is empty, as expected. But the `modularity` vector is not.

**Comment by szhorvat** [comment]
The problem was that when the graph is disconnected, both `modularity` and `merges` should be shorter than usual, namely of length `vcount - component_count` and `vcount - component_count + 1`, respectively. However, `modularity` was not initialized to the correct size.

When weights are present, sometimes a zero modularity score was computed as a tiny negative number such as `-7.45058e-08`, due to roundoff errors. Then the `which_max` function returned an element (an exact zero) of the too-ling `modularity` vector which was effectively out of range.

This will be fixed in #2043.

@vtraag could this roundoff error be caused by your last fix to modularity calculations in walktrap, and if yes, do we need to care at all? Probably not.
